Utilities/Logging.cs

```diff
@@ -117,42 +120,178 @@ private static void log(string path, string method, string level, string data)
+                if (isUniqueMessage(message, thisData))
+                {
+                    RollbarLocator.RollbarInstance.Info(message, thisData);
+                    Info("Reporting unique data, anonymous ID " + RollbarLocator.RollbarInstance.Config.Person.Id + ": " + message + thisData);
+                }
+                else
+                {
+                    Warn(@"Unable to report message """ + message + @""". Invalid data type " + data.GetType());
+                }
+            }
+            catch (Exception)
+            {
+                // Nothing to do
+            }
+        }
+    }
+
+    public class _Rollbar
+    {
+        // Exception handling (configuration instructions are at https://github.com/rollbar/Rollbar.NET)
+        // We have a limited data plan, so before sending exceptions and other reports, we shall use the API to check that the item is unique
+        // The Rollbar API test console is available at https://rollbar.com/docs/api/test_console/.
+
+        const string rollbarReadToken = "66e63ff290854a75b8b4c3263f084db6";
+        const string rollbarWriteToken = "debe6e50f82d4e8c955d5efafa79c789";
+        private static bool filterMessages = true; // We are rate limited, so keep this set to true unless we have a good reason to do otherwise.
+
+        public static void configureRollbarExceptionHandling(bool beta, string uniqueId)
+        {
+            var config = new RollbarConfig(rollbarWriteToken)
+            {
+                Environment = beta ? "development" : "production",
+                ScrubFields = new string[] // Scrub these fields from the reported data
+                {
+                    "Commander", "apiKey", "commanderName"
+                },
+                // Identify each EDDI configuration by a unique ID, or by "Commander" if a unique ID isn't available.
+                Person = new Rollbar.DTOs.Person(uniqueId),
+                // Set server info
+                Server = new Rollbar.DTOs.Server
+                {
+                    CodeVersion = Constants.EDDI_VERSION,
+                    Root = "/"
+                },
+            };
+            RollbarLocator.RollbarInstance.Configure(config);
+            // Send unhandled exceptions
+            AppDomain.CurrentDomain.UnhandledException += (sender, args) =>
+            {
+                Exception exception = args.ExceptionObject as Exception;
 
-                string body = @"{""message"":""" + message + @""", ""version"":""" + Constants.EDDI_VERSION + @""", ""json"":" + data + @"}";
-                Thread thread = new Thread(() =>
+                Dictionary<string, object> trace = new Dictionary<string, object>();
+                if (exception.StackTrace != null)
                 {
-                    try
+                    trace.Add("StackTrace", exception.StackTrace != null ? exception.StackTrace : "StackTrace not available");
```

Comment:
Due to the outer test, the else condition will never fire here. I'll rework it with the ?? operator